Fail during install: NoSuchAlgorithmException: unrecognized algorithm name

**OS version:** Windows 7 64-bit
**QZ Tray version:** (TBD)
**Java version:** (TBD)

Possibly related: https://bugs.java.com/bugdatabase/view_bug.do?bug_id=JDK-8266261

```
openjdk version "1.8.0_292"
OpenJDK Runtime Environment (AdoptOpenJDK)(build 1.8.0_292-b10)
OpenJDK Client VM (AdoptOpenJDK)(build 25.292-b10, mixed mode)
Output folder: C:\Program Files\QZ Tray
[...]
[ERROR] 2021-05-26 14:21:45,820 @ qz.utils.ArgParser:?
        Installation step CERTGEN failed
java.security.KeyStoreException: Key protection  algorithm not found: java.security.UnrecoverableKeyException: Encrypt Private Key failed: unrecognized algorithm name: PBEWithSHA1AndDESede
        at sun.security.pkcs12.PKCS12KeyStore.setKeyEntry(PKCS12KeyStore.java:677)
        at sun.security.pkcs12.PKCS12KeyStore.engineSetKeyEntry(PKCS12KeyStore.java:577)
        at java.security.KeyStore.setKeyEntry(KeyStore.java:1140)
        at qz.installer.certificate.CertificateManager.createKeyStore(Unknown Source)
        at qz.installer.certificate.CertificateManager.<init>(Unknown Source)
        at qz.installer.Installer.certGen(Unknown Source)
        at qz.utils.ArgParser.processInstallerArgs(Unknown Source)
        at qz.utils.ArgParser.intercept(Unknown Source)
        at qz.ws.PrintSocketServer.main(Unknown Source)
Caused by: java.security.UnrecoverableKeyException: Encrypt Private Key failed: unrecognized algorithm name: PBEWithSHA1AndDESede
        at sun.security.pkcs12.PKCS12KeyStore.encryptPrivateKey(PKCS12KeyStore.java:921)
        at sun.security.pkcs12.PKCS12KeyStore.setKeyEntry(PKCS12KeyStore.java:614)
        ... 8 more
Caused by: java.security.NoSuchAlgorithmException: unrecognized algorithm name: PBEWithSHA1AndDESede
        at sun.security.x509.AlgorithmId.get(AlgorithmId.java:448)
        at sun.security.pkcs12.PKCS12KeyStore.mapPBEAlgorithmToOID(PKCS12KeyStore.java:938)
        at sun.security.pkcs12.PKCS12KeyStore.encryptPrivateKey(PKCS12KeyStore.java:895)
        ... 9 more
Installation failed during certgen step.  Please check log for details.
```

![Capture135](https://user-images.githubusercontent.com/6345473/119508785-07b73000-bd3e-11eb-8fa4-73c7d715f01b.JPG)

--- a/src/qz/installer/certificate/CertificateManager.java
+++ b/src/qz/installer/certificate/CertificateManager.java
@@ -45,6 +45,11 @@ import static qz.installer.certificate.KeyPairWrapper.Type.*;
  * Stores and maintains reading and writing of certificate related files
  */
 public class CertificateManager {
+    static {
+        // Workaround for JDK-8266929
+        // See also https://github.com/qzind/tray/issues/814
+        SystemUtilities.clearAlgorithms();
+    }
     private static final Logger log = LoggerFactory.getLogger(CertificateManager.class);
 
     public static String DEFAULT_KEYSTORE_FORMAT = "PKCS12";

--- a/src/qz/utils/SystemUtilities.java
+++ b/src/qz/utils/SystemUtilities.java
@@ -537,4 +537,51 @@ public class SystemUtilities {
         }
         return hasMonocle;
     }
+
+    public static Version[] JDK_8266929 = {
+            Version.valueOf("11.0.11"),
+            Version.valueOf("1.8.0+291"),
+            Version.valueOf("1.8.0+292")
+    };
+
+    /**
+     * Fixes JDK-8266929 by clearing the oidTable
+     * See also: https://github.com/qzind/tray/issues/814
+     */
+    public static void clearAlgorithms() {
+        boolean needsPatch = false;
+        for(Version affected : JDK_8266929) {
+            if(affected.getMajorVersion() == 1) {
+                // Java 1.8 honors build/update information
+                if(affected.compareWithBuildsTo(Constants.JAVA_VERSION) == 0) {
+                    needsPatch = true;
+                }
+            } else if (affected.compareTo(Constants.JAVA_VERSION) == 0) {
+                // Java 9.0+ ignores build/update information
+                needsPatch = true;
+            }
+        }
+        if(!needsPatch) {
+            log.debug("Skipping JDK-8266929 patch for {}", Constants.JAVA_VERSION);
+            return;
+        }
+        try {
+            log.info("Applying JDK-8266929 patch");
+            Class<?> algorithmIdClass = Class.forName("sun.security.x509.AlgorithmId");
+            java.lang.reflect.Field oidTableField = algorithmIdClass.getDeclaredField("oidTable");
+            oidTableField.setAccessible(true);
+            // Set oidTable to null
+            oidTableField.set(algorithmIdClass, null);
+            // Java 1.8
+            if(Constants.JAVA_VERSION.getMajorVersion() == 1) {
+                java.lang.reflect.Field initOidTableField = algorithmIdClass.getDeclaredField("initOidTable");
+                initOidTableField.setAccessible(true);
+                // Set init flag back to false
+                initOidTableField.set(algorithmIdClass, false);
+            }
+            log.info("Successfully applied JDK-8266929 patch");
+        } catch (Exception e) {
+            log.warn("Unable to apply JDK-8266929 patch.  Some algorithms may fail.", e);
+        }
+    }
 }

--- a/src/qz/utils/SystemUtilities.java
+++ b/src/qz/utils/SystemUtilities.java
@@ -538,7 +538,7 @@ public class SystemUtilities {
         return hasMonocle;
     }
 
-    public static Version[] JDK_8266929 = {
+    public static final Version[] JDK_8266929_VERSIONS = {
             Version.valueOf("11.0.11"),
             Version.valueOf("1.8.0+291"),
             Version.valueOf("1.8.0+292")
@@ -550,7 +550,7 @@ public class SystemUtilities {
      */
     public static void clearAlgorithms() {
         boolean needsPatch = false;
-        for(Version affected : JDK_8266929) {
+        for(Version affected : JDK_8266929_VERSIONS) {
             if(affected.getMajorVersion() == 1) {
                 // Java 1.8 honors build/update information
                 if(affected.compareWithBuildsTo(Constants.JAVA_VERSION) == 0) {

